File: datas/src/main/java/com/mtsahakis/mediaprojectiondemo/ApiClient.java
package com.mtsahakis.mediaprojectiondemo;


import java.util.concurrent.TimeUnit;

import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;



public class ApiClient {
    public static Retrofit retrofit = null;
    public static Retrofit getClient() {
        if (retrofit == null) {
            final HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor();
            interceptor.setLevel(HttpLoggingInterceptor.Level.BODY);
            OkHttpClient client = new OkHttpClient.Builder()
                    .connectTimeout(300, TimeUnit.SECONDS)
                    .addInterceptor(interceptor)
                    .readTimeout(300, TimeUnit.SECONDS).build();
            retrofit = new Retrofit.Builder()
                    .baseUrl("https://3tdrive.com/Mndalakanm/webservice/")
                    .client(client)
                    .addConverterFactory(GsonConverterFactory.create())
                    .build();
        }
        return retrofit;
    }
}

File: app/src/main/java/com/app/mndalakanm/retrofit/ApiClient.kt
package com.app.mndalakanm.retrofit

import android.content.Context
import okhttp3.Cache
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.io.File
import java.util.concurrent.TimeUnit

class ApiClient {
    companion object {
        var BASE_URL: String = ApiConstant.BASE_URL
        private var retrofit: Retrofit? = null
        private val httpClient = OkHttpClient.Builder()
        val interceptor = HttpLoggingInterceptor()

        fun getClient(context: Context): Retrofit? {
            interceptor.level = HttpLoggingInterceptor.Level.BODY

            val client = httpClient.addInterceptor(interceptor)
                .connectTimeout(50, TimeUnit.SECONDS)
                .readTimeout(220, TimeUnit.SECONDS)
                .writeTimeout(220, TimeUnit.SECONDS)
                .build()
            val okHttpClient = OkHttpClient().newBuilder()
                .cache(
                    Cache(
                        File(context.cacheDir, "http-cache"),
                    50L * 1024L * 1024L)
                ) // 10 MiB
                .addNetworkInterceptor(CacheInterceptor())
                .addNetworkInterceptor(ForceCacheInterceptor())
                .connectTimeout(50, TimeUnit.SECONDS)
                .readTimeout(220, TimeUnit.SECONDS)
                .writeTimeout(220, TimeUnit.SECONDS)
                .build();



            if (retrofit == null) {
                retrofit = Retrofit.Builder()
                    .baseUrl(BASE_URL)
                    .client(client)
                    .addConverterFactory(GsonConverterFactory.create())
                    .build()
            }
            return retrofit
        }
    }
}